### PR TITLE
KS-129:  Implement modify DON

### PR DESCRIPTION
**Pull request by aiagentdev-1990**

### CODE REVIEW

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 1]
```diff
@@ -186,6 +192,11 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
   /// @param donId The ID of the removed DON
   event DONRemoved(uint32 donId);
 
+  /// @notice This event is emitted when a new DON is created
+  /// @param donId The ID of the updated DON
+  /// @param isPublic True if the updated DON is public
+  event DONUpdated(uint256 donId, bool isPublic);
```
Comment (aiagentdev-1990) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]: We currently emit this and `DONAdded` but I'm starting to think that it's redundant as we already emit `ConfigSet`.  The listener can derive whether or not the DON was added/updated using the `configCount`.

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 2]
```diff
@@ -638,4 +644,52 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
+    uint32 donId,
+    uint32 configCount,
+    bytes32[] calldata nodes,
+    CapabilityConfiguration[] calldata capabilityConfigurations,
+    bool isPublic
+  ) internal {
+    s_dons[donId].isPublic = isPublic;
+
+    DONCapabilityConfig storage donCapabilityConfig = s_dons[donId].config[configCount];
+    for (uint256 i; i < nodes.length; ++i) {
```
Comment (aiagentdev-1990) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol:648]: Thinking a bit on how to optimize this.  Currently the code blindly updates the DON's nodes and supported capability configs even if the function parameters are the same as what's already in storage.  One idea I had was to store the hash of `nodes` and `capabilityConfigs` and compare them against the hash of the params passed in.

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol (hunk 1 above)
Comment (DeividasK) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol]: I agree, this is a great idea! Should we remove DON... events and keep only `event DONConfigSet(uint32 donId, uint32 configCount);` event? configCount 0 (removed), 1 (new), 2+ (update) can be used to derive actions.

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol [hunk 3]
```diff
@@ -550,46 +561,34 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
...
-          revert NodeDoesNotSupportCapability(nodeP2PId, capabilityId);
-      }
-
-      donCapabilityConfig.capabilityIds.push(capabilityId);
-      donCapabilityConfig.capabilityConfigs[capabilityId] = configuration.config;
-    }
-
-    s_dons[id].configCount = configCount;
+    uint32 configCount = 1; // Initialize config count to start at 1
+    _setDONConfig(id, configCount, nodes, capabilityConfigurations, isPublic);
```
Comment (DeividasK) [inline contracts/src/v0.8/keystone/CapabilityRegistry.sol]:
```suggestion
    _setDONConfig(id, 1, nodes, capabilityConfigurations, isPublic);
```

File: contracts/src/v0.8/keystone/CapabilityRegistry.sol (hunk 2 above)
Comment (DeividasK) [reply contracts/src/v0.8/keystone/CapabilityRegistry.sol:648]: Create a ticket to test this out during the gas optimization stage?

**Review by DeividasK — APPROVED** [review]

**Pull request merged**